### ISSUE TITLE
Local headless service with clusterIP field set as none still requires ports field to be specified 

### ISSUE THREAD

**Issue by crimsonfaith91** [kubernetes/kubernetes#62440]
**What happened**:
When creating a headless service **locally** with `clusterIP` set as `None`, `ports` field has to be specified. Otherwise, endpoint controller will not be able to create an endpoint for the headless service successfully. This is because port number of to-be-created endpoint is defaulted to 0 when the `ports` field is not specified, which results in an error (0 is a reserved number not available for TCP and UDP usage).

This [tutorial](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods-hostname-and-subdomain-fields) explicitly suggests that the `ports` field is optional, which can be confusing to those unfamiliar with kubernetes networking.

**What you expected to happen**:
Port number of to-be-created endpoint should be defaulted to any value from 1 to 65535, not 0. Apart from this, the tutorial should not suggest the `ports` field as optional until the defaulting is fixed.

**How to reproduce it (as minimally and precisely as possible)**:
1. `kubectl create -f headless-service.yaml`
```
apiVersion: v1
kind: Service
metadata:
  name: default-subdomain
spec:
  selector:
    name: busybox
  clusterIP: None
```
2. `kubectl create -f busybox.yaml`: an [example](https://raw.githubusercontent.com/kubernetes/website/master/docs/tasks/administer-cluster/busybox.yaml) to refer, but remember to specify `subdomain` field
3. look into controller manager log and search for errors related to endpoint creation for the headless service

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0-dirty", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"dirty", BuildDate:"2018-04-12T21:29:37Z", GoVersion:"go1.10.1", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: local cluster
- OS (e.g. from /etc/os-release): `Debian GNU/Linux`

/cc @MrHohn @foxish

**Comment by MrHohn** [comment]
In endpoints controller we have explicit logic allowing headless service not to have ports:
https://github.com/kubernetes/kubernetes/blob/48388fec7eaad4ac8d84fbe20673ffacf41964a1/pkg/controller/endpoint/endpoints_controller.go#L454-L460

Though confirmed creating a headless service with no ports defined hit the exact same error:
```
endpoints_controller.go:375] Error syncing endpoints for service "default/XXX": Endpoints "XXX" is invalid: subsets[0].ports[0].port: Invalid value: 0: must be between 1 and 65535, inclusive
```

Will investigate more to see if this is a recent regression on endpoint API validation.

**Comment by MrHohn** [comment]
Ref #55158, which reported the same bug.

**Comment by crimsonfaith91** [comment]
@MrHohn Thanks for looking at the issue! Tell me if I can be of any help. :)

**Comment by MrHohn** [comment]
@crimsonfaith91 Sure thing :) I have a fix opened #62497.
/assign
